rust/common/error.rs

```diff
@@ -140,54 +100,151 @@ macro_rules! max {
     ($x:literal, $($xs:literal),*) => (max($x, max!($($xs),*)));
 }
 
+macro_rules! format_message {
+    ($self:ident, $error_name:ident, $body:literal, ) => {
+        format!($body)
+    };
+    ($self:ident, $error_name:ident, $body:literal, $t1:ty) => {
+        if let Self::$error_name(a) = &$self {
+            format!($body, a)
+        } else {
+            unreachable!()
+        }
```

Comment:
Because of Rust macro hygiene we have to resort to this double checking. In C-like macros we'd generate a full match arm, or generate bindings in the pattern and use them in the match result, but neither of those is allowed in Rust.